src/System.IO.FileSystem/src/System/IO/DosMatcher.cs

```diff
@@ -0,0 +1,307 @@
+{
+    internal static class DosMatcher
+    {
+        // [MS - FSA] 2.1.4.4 Algorithm for Determining if a FileName Is in an Expression
+        // https://msdn.microsoft.com/en-us/library/ff469270.aspx
+        private static readonly char[] s_wildcardChars =
+        {
+            '\"', '<', '>', '*', '?'
+        };
+
+        /// <summary>
+        /// Change '*' and '?' to '&lt;', '&gt;' and '"' to match Win32 behavior. For compatibility, Windows
+        /// changes some wildcards to provide a closer match to historical DOS 8.3 filename matching.
+        /// </summary>
+        public unsafe static string TranslateExpression(string expression)
+        {
+            if (string.IsNullOrEmpty(expression) || expression == "*.*")
+                return "*";
+
+            StringBuilder sb = StringBuilderCache.Acquire(expression.Length);
+            int length = expression.Length;
+            for (int i = 0; i < length; i++)
+            {
+                char c = expression[i];
+                switch (c)
+                {
+                    case '.':
+                        if (i > 1 && i == length - 1 && expression[i - 1] == '*')
+                        {
+                            sb.Length--;
+                            sb.Append('<'); // DOS_STAR (ends in *.)
+                        }
+                        else if (i < length - 1 && (expression[i + 1] == '?' || expression[i + 1] == '*'))
+                        {
+                            sb.Append('\"'); // DOS_DOT
+                        }
+                        else
+                        {
+                            sb.Append('.');
+                        }
+                        break;
+                    case '?':
+                        sb.Append('>'); // DOS_QM
+                        break;
+                    default:
+                        sb.Append(c);
+                        break;
+                }
+            }
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        /// <summary>
+        /// Return true if the given expression matches the given name.
+        /// </summary>
+        /// <param name="expression">The expression to match with, such as "*.foo".</param>
+        /// <param name="name">The name to check against the expression.</param>
+        /// <param name="ignoreCase">True to ignore case (default).</param>
+        /// <remarks>
+        /// This is based off of System.IO.PatternMatcher used in FileSystemWatcher, which is based off
+        /// of RtlIsNameInExpression, which defines the rules for matching DOS wildcards ('*', '?', '&lt;', '&gt;', '"').
+        /// 
+        /// Like PatternMatcher, matching will not line up with Win32 behavior unless you transform the expression
+        /// using <see cref="TranslateExpression(string)"/>
+        /// </remarks>
+        public unsafe static bool MatchPattern(string expression, ReadOnlySpan<char> name, bool ignoreCase = true)
```

Comment:
This will need good tests, I don't think I see good tests for the PatternMatcher to borrow, you might want to test against this internal method to get good coverage.

Reply:
I'm not going to try to review it carefully, I think tests would be more useful validation...

Reply:
I have a bunch. I need to get them ported.

Reply:
Note that we do have a lot of tests for this that hit it at a higher level. (That I specifically wrote based on the lower level tests.)